rayon-core/src/sleep/mod.rs

```diff
@@ -20,14 +37,51 @@ const ROUNDS_UNTIL_SLEEPY: usize = 32;
 const ROUNDS_UNTIL_ASLEEP: usize = 64;
 
 impl Sleep {
-    pub fn new() -> Sleep {
+    pub fn new(worker_count: usize) -> Sleep {
         Sleep {
             state: AtomicUsize::new(AWAKE),
-            data: Mutex::new(()),
+            data: Mutex::new(SleepData {
+                worker_count,
+                active_threads: worker_count,
+                blocked_threads: 0,
+                external_waiters: 0,
+            }),
             tickle: Condvar::new(),
         }
     }
 
+    #[inline]
+    pub fn add_external_waiter(&self) {
+        let mut data = self.data.lock().unwrap();
+        data.external_waiters += 1;
+    }
+
+    #[inline]
+    pub fn sub_external_waiter(&self) {
+        let mut data = self.data.lock().unwrap();
+        data.external_waiters -= 1;
+    }
+
+    #[inline]
+    pub fn block(&self, deadlock_handler: &Option<Box<DeadlockHandler>>) {
+        let mut data = self.data.lock().unwrap();
+        debug_assert!(data.blocked_threads < data.worker_count);
+        debug_assert!(data.active_threads > 0);
+        data.active_threads -= 1;
+        data.blocked_threads += 1;
+
+        data.deadlock_check(deadlock_handler);
+    }
+
+    #[inline]
+    pub fn unblock(&self) {
```

Comment:
Same here, `mark_unblocked()` seems clearer to me.

Reply:
Renamed.